core/src/main/java/software/amazon/awssdk/core/internal/http/pipeline/stages/MakeAsyncHttpRequestStage.java

```diff
@@ -197,4 +202,42 @@ public void exceptionOccurred(Throwable throwable) {
         }
 
     }
+
+    /**
+     * An interface similar to {@link CompletableFuture} that may or may not dispatch completion of the future to an executor
+     * service, depending on the client's configuration.
+     */
+    private class Completable {
+        private final CompletableFuture<Response<OutputT>> completableFuture = new CompletableFuture<>();
+
+        public void complete(Response<OutputT> result) {
+            switch (futureCompletionThread) {
+                case EVENT_LOOP:
+                    completableFuture.complete(result);
+                    break;
+                case EXECUTOR_SERVICE:
```

Comment:
Instead of the enum can we just wire up an ExecutorService that runs things on the current thread?

Reply:
I tried that originally, but was worried it was going to be confusing to someone reading the code in the future, because it pushed control of the threading model all the way up into the client builder. I figured we could always push it higher up if we wanted to in the future since it's entirely local to this module. LMK if you disagree. It's easy to switch it back.

Reply:
I think technically that violates the contract of the interface though since `ExecutorService`s are supposed to handle tasks asynchronously.

I like the idea of exposing something like `FutureCompletionStrategy` though

Reply:
I'd say it belongs in the client builder since that's where everything is wired up and defaults are chosen. I don't think running it synchronously breaks the contract of executor service, Guava for example exposes a similar implementation. In fact the documentation for Executor mentions the sync use case explicitly (ExecutorService mentions async only but perhaps we should be using Executor anyways unless we need the additional methods).

Reply:
+1 for `Executor`